lib/auth/auth.js

```diff
@@ -32,41 +28,64 @@ auth.doAuth = (request, log, cb, awsService) => {
         }
-    } else if (request.query.Signature) {
+        log.warn('missing authorization security header');
+        return { err: errors.MissingSecurityHeader };
+    } else if (data.Signature) {
         // Check whether signature is in query string
         log.trace('authenticating request with auth v2 using query string');
-        authV2.queryAuthCheck.check(request, log, cb);
-    } else if (request.query['X-Amz-Algorithm']) {
+        return authV2.queryAuthCheck.check(request, log, data);
+    } else if (data['X-Amz-Algorithm']) {
         log.debug('authenticating request with Auth v4 using query string');
-        authV4.queryAuthCheck.check(request, log, cb);
-    } else {
-        // If no auth information is provided in request, then
-        // user is part of 'All Users Group' so send back this
-        // group as the canonicalID
-        log.trace('No authentication provided. User identified as public');
-        const authInfo = new AuthInfo({ canonicalID: constants.publicId });
-        return cb(null, authInfo);
+        return authV4.queryAuthCheck.check(request, log, data);
+    }
+    // If no auth information is provided in request, then
+    // user is part of 'All Users Group' so send back this
+    // group as the canonicalID
+    log.trace('No authentication provided. User identified as public');
+    const authInfo = new AuthInfo({ canonicalID: constants.publicId });
+    return { err: null, data: authInfo };
+};
+
+auth.doAuth = (request, log, cb, awsService, data) => {
+    const res = auth.check(request, log, awsService, data);
+    if (res.err) {
+        return cb(res.err);
+    } else if (res.version === 2) {
+        return vault.authenticateV2Request(res.data.accessKey,
+            res.data.signatureFromRequest,
+            res.data.stringToSign, res.data.algo, log,
+            (err, authInfo) => {
+                if (err) {
+                    return cb(err);
+                }
+                return cb(null, authInfo);
+            });
+    } else if (res.version === 4) {
+        res.data.log = log;
+        return vault.authenticateV4Request(res.data, cb, awsService);
+    } else if (res.data instanceof AuthInfo) {
+        return cb(null, res.data);
     }
-    return undefined;
+    log.error('Cannot found authentification method', {
+        method: 'Arsenal.auth.doAuth',
+    });
+    return cb(errors.InternalError);
 };
 
 auth.generateV4Headers =
-    (request, data, accessKey, secretKeyValue) => {
+    (request, data, accessKey, secretKeyValue, awsService) => {
         Object.assign(request, { headers: {} });
         const amzDate = convertUTCtoISO8601(Date.now());
         // get date without time
         const scopeDate = amzDate.slice(0, amzDate.indexOf('T'));
         const signedHeaders = 'host;x-amz-date;x-amz-content-sha256';
         const region = 'us-east-1';
-        const service = 'iam';
+        const service = awsService || 'iam';
```

Comment:
It's not :) and it's more efficient for v8 in that way :) but i can change it :)